File: 0112-path-sum/0112-path-sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def hasPathSum(self, root, targetSum):
        """
        :type root: TreeNode
        :type targetSum: int
        :rtype: bool
        """
        if not root:
            return False #No nodes return false

        if not root.left and not root.right:
            return targetSum == root.val #if the root has no child nodes check if the root matches the target and return True or False accordingly

        left_has_path = self.hasPathSum(root.left, targetSum - root.val)
        right_has_path = self.hasPathSum(root.right, targetSum - root.val)

        return left_has_path or right_has_path

        

File: 0112-path-sum/0112-path-sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool HasPathSum(TreeNode root, int targetSum) {
        if (root == null)
        {
            return false;
        }
        if (root.left == null && root.right == null)
        {
            return targetSum == root.val;
        }
        bool leftHasPath = this.HasPathSum(root.left, targetSum - root.val);
        bool rightHasPath = this.HasPathSum(root.right, targetSum - root.val);
        
        return leftHasPath || rightHasPath;
    }
}